kolibri/core/assets/src/core-actions.js

```diff
@@ -54,9 +126,214 @@ function togglemodal(store, bool) {
+
+  /* Mark completion time if 100% progress reached */
+  const completedContent = originalProgress < 1 && summaryProgress === 1;
+  if (completedContent) {
+    store.dispatch('SET_LOGGING_COMPLETION_TIME', new Date());
+  }
+
+  /* Save models if needed */
+  if (forceSave || completedContent || progressThresholdMet) {
+    this.saveLogs(Kolibri);
+  }
+}
+
+
+/**
+ * Update the total time spent and end time stamps
+ * To be called periodically by interval timer
+ * Must be called after initContentSession
+ * @param {boolean} forceSave
+ */
+function updateTimeSpent(store, Kolibri, forceSave = false) {
+  /* Create aliases for logs */
+  const summaryLog = store.state.core.logging.summary;
+  const sessionLog = store.state.core.logging.session;
+
+  /* Calculate new times based on how much time has passed since last save */
+  const sessionTime = intervalTimer.getNewTimeElapsed() + sessionLog.time_spent;
+  const summaryTime = (summaryLog.id) ?
+    sessionTime + summaryLog.time_spent_before_current_session : 0;
+
+  /* Update the logging state with new timing information */
+  store.dispatch('SET_LOGGING_TIME', sessionTime, summaryTime, new Date());
+
+  /* Determine if time threshold has been met */
+  const timeThresholdMet = sessionLog.time_spent -
+    sessionLog.total_time_at_last_save >= timeThreshold;
+
+  /* Save models if needed */
+  if (forceSave || timeThresholdMet) {
+    this.saveLogs(Kolibri);
+  }
+}
+
+
+/**
+ * Do a PATCH to update existing logging models
+ * Must be called after initContentSession
+ */
+function saveLogs(store, Kolibri) {
+  const ContentSessionLogResource = Kolibri.resources.ContentSessionLogResource;
+  const ContentSummaryLogResource = Kolibri.resources.ContentSummaryLogResource;
+  /* Create aliases for logs */
+  const summaryLog = store.state.core.logging.summary;
+  const sessionLog = store.state.core.logging.session;
+
+  /* If a session model exists, save it with updated values */
+  if (sessionLog.id) {
+    const sessionModel = ContentSessionLogResource.getModel(sessionLog.id);
+    sessionModel.save(_contentSessionModel(store)).then((model) => {
+      /* Reset values used for threshold checking */
+      sessionLog.total_time_at_last_save = model.time_spent;
+      sessionLog.progress_at_last_save = model.progress;
+    }).catch((error) => {
+      store.dispatch('SET_ERROR', JSON.stringify(error, null, '\t'));
+    });
+  }
+
+  /* If a summary model exists, save it with updated values */
+  if (summaryLog.id) {
+    const summaryModel = ContentSummaryLogResource.getModel(summaryLog.id);
+    summaryModel.save(_contentSummaryModel(store)).then((model) => {
+      /* PLACEHOLDER */
+    }).catch((error) => {
+      store.dispatch('SET_ERROR', JSON.stringify(error, null, '\t'));
+    });
+  }
+}
+
+
+/**
+ * Start interval timer and set start time
+ * @param {int} interval
+ */
+function startTrackingProgress(store, Kolibri, interval = intervalTime) {
+  const self = this;
```

Comment:
I am a little suprised that this works, as I don't think `this` refers to the Vue component that the action is bound to. Does this work?

You may not even need to do it, as you have a direct reference to the `updateTimeSpent` action because it is in this file :)